### PR TITLE
[AD-203] Fix where clause

### PR DESCRIPTION
**Pull request by mitchell-elholm**
### Summary

AD-203 Fix for where clause

### Description

Fixes multiple issues related to where clause:

- Multiple columns in a comparison, eg fieldA > fieldB
- Boolean literal, eg. fieldA = FALSE
- Nested OR condition eg. WHERE (A OR B) AND (C OR D)

It accomplishes this by re-using the aggregation framework to add a boolean field representing the value of the query in the where clause. This is then used to determine whether the document should be filtered out. The Translator previously used in DocumentDbFilter has been commented out as it may still prove useful later.

Does not fix [AD-234 Fix where clause with false comparison](https://bitquill.atlassian.net/browse/AD-234).

### Related Issue

https://bitquill.atlassian.net/browse/AD-203

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java]: Would it make sense to wrap the value in NumberLong() instead of $literal? https://docs.mongodb.com/manual/core/shell-types/#numberlong

**Review by birschick-bq — COMMENTED** [review]
Merge the `develop` branch into this feature branch to remove already merged code.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java]: Changed to use $numberLong instead.

**Comment by mitchell-elholm** [comment]
Will wait for confirmation on $literal support before merging.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:40]: the name seems pretty collision proof to me honestly 😅  Not sure if we should add a check or something that this name isn't in the metadata or have a way of randomly generating the name each time

Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:40]: That was definitely the goal, to use a name that definitely won't be used by accident. We could probably implement something to randomly generate a name, and then check the metadata that it isn't there, and re-generate a new random name if it is. I just put a long string of letters and numbers assuming nobody would accidentally try and use that because it was easier.

Comment (alexey-temnikov) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:40]: Usually UUID provide quite reliable collision proof, adding a prefix should do the thing.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:93]: As discussed - add comment with the link to the reference for the ticket and add comment that code was commented out temporary, and waiting to be restored as soon as $literal support is added.

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementTest.java
Comment (alexey-temnikov) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementTest.java:1531]: Add unit tests for WHERE equality conditions, e.g.
```
SELECT "tdvt"."Products"."name", "tdvt"."Products"."available"  FROM "tdvt"."Products" where "tdvt"."Products"."available" = 'false';
```

Also let's validate each data type is supported.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:93]: Added comment.

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementTest.java
Comment (mitchell-elholm) [reply src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementTest.java:1531]: Added test with additional types and equality condition.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:40]: Maybe we can make a ticket to possibly revisit using UUID generation with this

**Review by andiemontoyeah — APPROVED** [review]

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:133]: nit: did you accidentally change `*` to `o`? This happens in a few places.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java]: nit: you may want to make this use the constant instead of string value.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbFilter.java:133]: It wasn't an accident, I did that to make it comment out the entire Translator class, rather than ending the comment block at the start of one of the existing comments. Not sure if this is a weird thing to do, but I think I saw this done elsewhere in the project.

**Review by birschick-bq — APPROVED** [review]

**Pull request merged**
